### ISSUE TITLE
Explore a validation for dynamic rules that encourages them to go from most to least specific

### ISSUE THREAD

**Issue by ross** [octodns/octodns#989]
https://github.com/octodns/octodns-ns1/issues/36 is tracking some potential issues with dynamic rules and geo's with NS1, but while looking into it I realized the example config wasn't ordered most to least specific which would generally be the "best practice." 

I'm not even sure that's said anywhere. If not it should be in the dynamic doc. It'd also be worth exploring a validation check for the ordering of dynamic rules to check their order. The validation can always be disabled if someone has a reason to do otherwise.

- [x] Make sure the best practices are documented
- [x] Explore a validation that checks for reusing targeted locs
- [x] Explore a validation that checks for targeting a more specific loc inside of a previous one
- [x] Explore a validation that verifies there's a "catch-all" rule at the end

/cc https://github.com/octodns/octodns-ns1/issues/36

**Comment by viranch** [comment]
Talking of best practice, I would also make a case for automatically appending a catch-all rule (in absence of one) which points to the top-level `value`/`values`.

**Comment by ross** [comment]
> Talking of best practice, I would also make a case for automatically appending a catch-all rule (in absence of one) which points to the top-level `value`/`values`.

Feel like that'd best be a warning/error as there aren't any other cases I can think of where octoDNS modifies things in that sort of way.

**Comment by ross** [comment]
2 of the 4 checkboxes are done in https://github.com/octodns/octodns/pull/991, will pick up the other 2 in there later/tomorrow.

**Comment by istr** [comment]
I did some more research and thought over the discussion we had around the issues in the ns1 provider.

As a conclusion, I think that the concept of "sorted `rules`" does not make much sense, and is -- at best -- an implementation artifact. The _declarative semantics_ should reflect a configuration that is unambiguous and provides a straight "most specific rule wins" logic. Warnings generated if there are issues should reflect this logic as well.

Here are some reasons:

1. none of the currently implemented providers with `dynamic` facilities (apart from ns1) have a logic that would allow "ordered" rulesets
    - route53 explicitly selects the most specific target. "If you create separate records for overlapping geographic regions—for example, one record for North America and one for Canada—priority goes to the smallest geographic region."
      https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/routing-policy-geo.html
      Any kind of rule order here is pointless in the good case and misleading in bad cases.
     - ns1 takes the order of rules into account. However, tracking is _not natively supported_ (a workaround with annotations is used to track and sync the order)
     - constellix does not support ordered rules. Just like route53 the most specific rule "wins". Currently, the rules are sorted according to the pool name order.
       https://github.com/octodns/octodns-constellix/blob/244ed9447a140283ca7112ffefe26398ef840c55/octodns_constellix/__init__.py#L594
     - gcore sorts the `rules` according to the pool names as well
       https://github.com/octodns/octodns-gcore/blob/4235510fa8e6f04cf09dda3aa154dcf57c7719ad/octodns_gcore/__init__.py#L226
        gcore does not take rule ordering into account as per
        https://gcore.com/support/articles/360012179878/
     -  azure does not provide rule ordering, it will "start with the most specific rule" just as route53 and constellix, see
       https://learn.microsoft.com/en-us/azure/traffic-manager/traffic-manager-routing-methods#geographic-traffic-routing-method
    - edgecenter sorts the rules by pool name as well
      https://github.com/octodns/octodns-edgecenter/blob/9d9f25993a53715ad4efecb83d39918bc7902100/octodns_edgecenter/__init__.py#L238
      As per documentation it has a best match approach as well:
      https://support.edgecenter.ru/knowledge_base/item/257958?sid=57173 (Russian, need some translation tool)
2. it was already stated elsewhere that most providers do not allow for "reusing" pools in multiple rules.
    https://github.com/octodns/octodns/pull/991#issuecomment-1491018659
    To start with, that means that it is probably even pointless to have separate `rules` section to start with

So instead of trying to impose restrictions on the `rules` (that are in fact directly tied to the `pools` in most implementations) I would rather

- have a clear documentation and statement of the _intended semantics_ (without semantic meaning in `rules` order that will not be properly transported to most implementations anyway)
- consider to move the `rules` directly into the `pools`, given that this is a 1:1 relationship anyway; enforcing this relationship is currently cumbersome and would be much easier if the rules are integral part of the pools logic

**Comment by ross** [comment]
> As a conclusion, I think that the concept of "sorted `rules`" does not make much sense, and is -- at best -- an implementation artifact

I'm now 6 years removed from starting octoDNS and almost 5 from when things switched from `geo` to `dynamic` so a lot of super low level details, like the reasoning, are lost to time or at the very least faded.  At a high level the implementation is based on looking at all of the providers I had access to and seeing how they did things and then trying to synthesize a scheme that could be implemented on top of all of them with as much functionality & flexibility as possible. 

The specific requirement of having ordered rules may well have been a non-essential artifact, but it's unlikely to have been as a result of implementation as the design was discussed & sketched out extensively before it was coded. At the time I believe Dyn, NS1, and Route53 were the main providers I was looking at for `dynamic` support. Cloudflare (which is too different) was also considered.

I don't remember what lead to the rule ordering, but I do vaguely remember trying to start out with rules being a dictionary similar to pools and that running into issues. Maybe whatever those issues were lead to the ordering requirement. It's also important to consider that we may have been taking into account potential support for things beyond geo-location, as @viranch's recent PR will start to introduce. 

My best guess is that some provider(s) applied things in a strict order and thus couldn't handle the case where 2 rules had opposing specific & broad geos, a la https://github.com/octodns/octodns/pull/991#issuecomment-1491018659.  Assuming that's the case then imposing the order on things (and the best practice/validation in https://github.com/octodns/octodns/pull/991) will allow all the providers to behave the same at the expense losing the ability to have opposing specific rules.

I haven't been in the code in 4-5 years now, but Dyn may have had explicit rule ordering as a hard requirement which may have weighed on this direction. It's `DSFRuleset` appears to have an ordering field. 

> ...

TL;DR it may well be the case that most/all of the active providers at this point would support unordered rules, but Dyn may not have and/or there may have been thinking/planning/allowing for supporting other targeting methods. There are a lot of trade offs in dynamic records where things are more constrained than (some providers) would allow them to be. This is definitely a case of that.



> * have a clear documentation and statement of the _intended semantics_ (without semantic meaning in `rules` order that will not be properly transported to most implementations anyway)

Definitely trying to get things to be more clear/explicit with the combination of validations & documentation in https://github.com/octodns/octodns/pull/991#issuecomment-1491018659. 

I think the rule ordering & best practices :ship: has for better or worse sailed at this point given that it's been in the `dynamic` the implementation for 5 years. The new validations, like all validations, can be disabled with `lenient` for anyone who prefers to treat rules as an unordered pool. That should even allow the opposing targeting that some providers can support, but others cannot. 

That's the crux of things here. The schema along with defaults, best practices, and validations are intended to be something that can be implemented pretty much equally across all providers. And `lenient` is the "I know what I'm doing" flag that says actually I want you to do this and I'll take on making sure that the provider(s) I use will do the right thing with it.

> * consider to move the `rules` directly into the `pools`, given that this is a 1:1 relationship anyway; enforcing this relationship is currently cumbersome and would be much easier if the rules are integral part of the pools logic

This actually isn't the case. It's close, but not quite. The catch-all rule can re-use a pool that was previously used in a `geos` rule. This is essentially for cases where you're saying things like: I want all of NA to go to `iad`, EU to `fra`, ... and if you can't geo-locate something or it's not covered explicitly by a rule above send it to `iad`. 

I don't think having these validations prevents people from doing things the want/can do, but it's a guardrail that keeps people on the best practices road when they don't want to/aren't up for reading all the code, figuring out the details, and making sure that what they expect to happen will actually happen.

**Comment by istr** [comment]
@ross,  I hope you didn't get me wrong: this was not meant to be a criticism of what the implementation looks like now, nor how it came to be, at all.

octodns works _wonderfully_ and has evolved magnificently over the last few years.

The comment was more an invitation to step back and rethink how to make the next iteration simpler and thus better.

That's the way it is with all reasonably successful projects: you start with a small problem that you want to solve. Then it grows and at some point things have changed so much that you have to look at it again and maybe rework some things.

I would continue to focus on what octodns does well and why it is superior to other solutions (for example terraform modules): to carry the same crystal clear semantics from a single declarative configuration source across many different providers.

The crux and the really hard part from my point of view is rather to build a good migration path to the next, improved iteration. And also in this respect you have done a very good job with octodns so far.

**Comment by ross** [comment]
Implemented in https://github.com/octodns/octodns/pull/991. So this is resolved (for now)
